pkg/disttask/framework/dispatcher/dispatcher_manager.go

```diff
@@ -162,71 +152,67 @@ func (dm *Manager) dispatchTaskLoop() {
 			logutil.BgLogger().Info("dispatch task loop exits", zap.Error(dm.ctx.Err()), zap.Int64("interval", int64(checkTaskRunningInterval)/1000000))
 			return
 		case <-ticker.C:
-			cnt := dm.getRunningTaskCnt()
-			if dm.checkConcurrencyOverflow(cnt) {
-				break
-			}
+		}
 
-			// TODO: Consider getting these tasks, in addition to the task being worked on..
-			tasks, err := dm.taskMgr.GetTasksInStates(
-				dm.ctx,
-				proto.TaskStatePending,
-				proto.TaskStateRunning,
-				proto.TaskStateReverting,
-				proto.TaskStateCancelling,
-				proto.TaskStateResuming,
-			)
-			if err != nil {
-				logutil.BgLogger().Warn("get unfinished(pending, running, reverting, cancelling, resuming) tasks failed", zap.Error(err))
-				break
+		taskCnt := dm.getDispatcherCount()
+		if taskCnt >= proto.MaxConcurrentTask {
+			logutil.BgLogger().Info("dispatched tasks reached limit",
+				zap.Int("current", taskCnt), zap.Int("max", proto.MaxConcurrentTask))
+			continue
+		}
+
+		tasks, err := dm.taskMgr.GetTopUnfinishedTasks(dm.ctx)
+		if err != nil {
+			logutil.BgLogger().Warn("get unfinished tasks failed", zap.Error(err))
+			continue
+		}
+
+		dispatchableTasks := make([]*proto.Task, 0, len(tasks))
+		for _, task := range tasks {
+			if dm.hasDispatcher(task.ID) {
+				continue
 			}
+			// we check it before start dispatcher, so no need to check it again.
+			// see startDispatcher.
+			// this should not happen normally, unless user modify system table
+			// directly.
+			if getDispatcherFactory(task.Type) == nil {
```

Comment:
Which system table affects this behavior?

Reply:
suppose user change task.Type to some invalid value